### PR TITLE
enhance env printing + add  `--capture_output`  option

### PR DESCRIPTION
**Pull request by hjanuschka**
this one refactors the :env  a little bit, and adds a global parameter `--trace` - to capture all fastlane output.

so if users having trouble using fastlane they just run their failing command with the `--trace` 

e.g.:

```
fastlane demo --capture_output

```

sample:
### Trace Info

Command Used: `demo`
<details><summary>

Output/Log</summary>

```

[16:56:31]: Driving the lane 'demo' 🚀
[16:56:32]: fastlane.tools finished successfully 🎉


```

</details>

<details><summary>

✅ fastlane environment ✅</summary>

### Stack

| Key | Value |
| --- | --- |
| OS | 10.12 |
| Ruby | 2.3.1 |
| Bundler? | true |
| Xcode Path | /Applications/Xcode.app/Contents/Developer/ |
| Xcode Version | 8.0 |
| Git | git version 2.8.4 (Apple Git-73) |
| Installation Source | /usr/local/lib/ruby/gems/2.3.0/bin/fastlane |
| Host | Mac OS X 10.12 (16A323) |
| Ruby Lib Dir | /usr/local/Cellar/ruby/2.3.1_2/lib |
| OpenSSL Version | OpenSSL 1.0.2j  26 Sep 2016 |
### fastlane files:

<details><summary>

`./fastlane/Fastfile`</summary>

``` ruby
lane :demo do
demo=1
end
```

</details>

**No Appfile found**
### fastlane gems

| Gem | Version | Update-Status |
| --- | --- | --- |
| credentials_manager | 0.16.2 | ✅ Up-To-Date |
| fastlane_core | 0.53.0 | ✅ Up-To-Date |
| spaceship | 0.36.1 | ✅ Up-To-Date |
| cert | 1.4.3 | ✅ Up-To-Date |
| deliver | 1.14.3 | ✅ Up-To-Date |
| frameit | 3.0.0 | ✅ Up-To-Date |
| gym | 1.11.3 | ✅ Up-To-Date |
| sigh | 1.11.2 | ✅ Up-To-Date |
| match | 0.9.0 | ✅ Up-To-Date |
| pem | 1.3.2 | ✅ Up-To-Date |
| pilot | 1.11.0 | ✅ Up-To-Date |
| produce | 1.2.1 | ✅ Up-To-Date |
| scan | 0.13.1 | ✅ Up-To-Date |
| screengrab | 0.5.5 | ✅ Up-To-Date |
| snapshot | 1.16.2 | ✅ Up-To-Date |
| supply | 0.7.1 | ✅ Up-To-Date |
| fastlane | 1.106.1 | ✅ Up-To-Date |
### Loaded fastlane plugins:

**No plugins Loaded***

<details><summary>

<b>Loaded gems</b></summary>

| Gem | Version |
| --- | --- |
| did_you_mean | 1.0.0 |
| bundler | 1.13.6 |
| io-console | 0.4.5 |
| rake | 11.3.0 |
| i18n | 0.7.0 |
| json | 1.8.3 |
| minitest | 5.9.1 |
| thread_safe | 0.3.5 |
| tzinfo | 1.2.2 |
| activesupport | 4.2.7.1 |
| addressable | 2.4.0 |
| ast | 2.3.0 |
| babosa | 1.0.2 |
| builder | 3.2.2 |
| byebug | 9.0.6 |
| colored | 1.2 |
| highline | 1.7.8 |
| commander | 4.4.0 |
| security | 0.1.3 |
| credentials_manager | 0.16.2 |
| excon | 0.54.0 |
| gh_inspector | 1.0.2 |
| multi_json | 1.12.1 |
| plist | 3.1.0 |
| rubyzip | 1.1.7 |
| terminal-table | 1.4.5 |
| fastlane_core | 0.53.0 |
| multipart-post | 2.0.0 |
| faraday | 0.9.2 |
| unf_ext | 0.0.7.2 |
| unf | 0.1.4 |
| domain_name | 0.5.20160826 |
| http-cookie | 1.0.3 |
| faraday-cookie_jar | 0.0.6 |
| faraday_middleware | 0.10.0 |
| fastimage | 1.6.8 |
| multi_xml | 0.5.5 |
| claide | 1.0.1 |
| cork | 0.2.0 |
| nap | 1.1.0 |
| open4 | 1.3.4 |
| claide-plugins | 0.9.2 |
| coderay | 1.1.1 |
| docile | 1.1.5 |
| simplecov-html | 0.10.0 |
| simplecov | 0.12.0 |
| tins | 1.12.0 |
| term-ansicolor | 1.4.0 |
| thor | 0.19.1 |
| coveralls | 0.8.15 |
| safe_yaml | 1.0.4 |
| crack | 0.4.3 |
| faraday-http-cache | 1.3.1 |
| git | 1.3.0 |
| sawyer | 0.7.0 |
| octokit | 4.4.0 |
| redcarpet | 3.3.4 |
| danger | 0.10.1 |
| diff-lcs | 1.2.5 |
| dotenv | 2.1.1 |
| fakefs | 0.8.1 |
| mini_magick | 4.5.1 |
| rouge | 1.11.1 |
| xcpretty | 0.2.4 |
| net-ssh | 3.2.0 |
| net-sftp | 2.1.2 |
| krausefx-shenzhen | 0.14.10 |
| slack-notifier | 1.5.1 |
| xcpretty-travis-formatter | 0.0.4 |
| jwt | 1.5.6 |
| little-plugger | 1.1.4 |
| logging | 2.1.0 |
| memoist | 0.15.0 |
| os | 0.9.6 |
| signet | 0.7.3 |
| googleauth | 0.5.1 |
| httpclient | 2.8.2.4 |
| hurley | 0.2 |
| mime-types | 1.25.1 |
| uber | 0.0.15 |
| representable | 2.3.0 |
| retriable | 2.1.0 |
| google-api-client | 0.9.19 |
| terminal-notifier | 1.7.1 |
| word_wrap | 1.0.0 |
| xcode-install | 2.0.7 |
| xcodeproj | 1.3.3 |
| method_source | 0.8.2 |
| parser | 2.3.1.4 |
| powerpack | 0.1.1 |
| slop | 3.6.0 |
| pry | 0.10.4 |
| pry-byebug | 3.4.0 |
| rainbow | 2.1.0 |
| rest-client | 1.6.9 |
| rspec-support | 3.1.2 |
| rspec-core | 3.1.7 |
| rspec-expectations | 3.1.2 |
| rspec-mocks | 3.1.3 |
| rspec | 3.1.0 |
| rspec_junit_formatter | 0.2.3 |
| ruby-progressbar | 1.8.1 |
| unicode-display_width | 1.1.1 |
| rubocop | 0.44.1 |
| webmock | 1.19.0 |
| yard | 0.8.7.6 |

</details>

_generated on:_ **2016-10-24**
</details>

### CODE REVIEW

**Review by milch — DISMISSED** [review]
Wow, I really like this change, thanks a lot! 👍

**Comment by milch** [comment]
Just one more thing before we can merge this: can you add the `--trace` global parameter to `fastlane/lib/fastlane/commands_generator.rb` so that it shows up when people run `fastlane --help`? Preferably with a short summary of what it does

File: fastlane/lib/fastlane/commands_generator.rb [hunk 1]
```diff
@@ -24,6 +32,15 @@ def self.start
     ensure
       FastlaneCore::UpdateChecker.show_update_status('fastlane', Fastlane::VERSION)
       Fastlane::PluginUpdateManager.show_update_status
+      if $trace
+        $trace_output = $stdout.string.gsub(/\e\[([;\d]+)?m/, '')
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/commands_generator.rb]: What is this for?

File: fastlane/lib/fastlane/commands_generator.rb [hunk 2]
```diff
@@ -24,6 +32,15 @@ def self.start
...
+        $stdout = STDOUT
+        $stderr = STDERR
+
+        require "fastlane/environment_printer"
+        Fastlane::EnvironmentPrinter.output
+
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/commands_generator.rb]: Please remove the newline here

**Review by KrauseFx — DISMISSED** [review]

File: fastlane/lib/fastlane/commands_generator.rb (hunk 1 above)
Comment (mfurtak) [inline fastlane/lib/fastlane/commands_generator.rb]: This looks like it removing ANSI color escape codes, but that's definitely worth documenting in a comment, or putting in a named method to make it clearer 👍

**Comment by mfurtak** [comment]
I haven't given this a try yet, but it seems cool 👍 

Can I suggest we pick a different name for this flag, maybe `--capture_output`? If I was guessing what something named `--trace` does, I'd guess either:

* forces display of a stack trace on error
* enables very verbose logging (sometimes called trace logging)

**Comment by hjanuschka** [comment]
@milch  - added the `global_option` description - so it shows up in `--help`

@mfurtak  - you are right about `--trace`  - i changed it to `--capture-output` - as the `_` in a optarg breaks my posix heart 😢  - hope this is ok

let me know if this PR is done, and many thx for your help and time

**Comment by mfurtak** [comment]
@hjanuschka I feel your pain when it comes to the `_` in the arg name. However, it's probably worth noting that all of _fastlane_'s long name args are done this way 😦  So, for consistency with long arg names everywhere else, I think I'd still suggest using the underscore.

**Comment by mfurtak** [comment]
@hjanuschka Have you seen `FASTLANE_DISABLE_COLORS`? Is it possible to set that env variable in an early enough place (based on your flag) that removes the need to strip it out later?

**Comment by hjanuschka** [comment]
@mfurtak  ok renamed it to `capture_output`  - i know that about the other args, everytime i run a action from commandline directly. i get like 4 heartattacks on how params are specified!

i even thought of refactoring it, but its a pain in the ass to  not break existing scripts!

**Comment by hjanuschka** [comment]
yeah saw the `FASTLANE_DISABLE_COLORS` - can we keep the explicit strip version just to be sure no ansi color is in it? - maybe some wired `puts` or anything else does coloring too.

File: fastlane/lib/fastlane/commands_generator.rb [hunk 3]
```diff
@@ -11,7 +11,15 @@ def self.start
       # since at this point we haven't yet loaded commander
       # however we do want to log verbose information in the PluginManager
       $verbose = true if ARGV.include?("--verbose")
-
+      $capture_output = true if ARGV.include?("--capture-output")
```
Comment (mfurtak) [inline fastlane/lib/fastlane/commands_generator.rb]: Change to `--capture_output`

**Review by milch — APPROVED** [review]
Ready once the proposed change by @mfurtak is done 🚀

File: fastlane/lib/fastlane/commands_generator.rb (hunk 3 above)
Comment (hjanuschka) [inline fastlane/lib/fastlane/commands_generator.rb]: god damn, sorry for wasting your time, i can't get this **_** into my head 🔫

**Comment by hjanuschka** [comment]
@milch @mfurtak  updated!

**Comment by milch** [comment]
Thanks 🚀

**Pull request merged**